source/common/router/config_impl.cc

```diff
@@ -383,6 +383,38 @@ RouteConstSharedPtr PathRouteEntryImpl::matches(const Http::HeaderMap& headers,
   return nullptr;
 }
 
+RegexRouteEntryImpl::RegexRouteEntryImpl(const VirtualHostImpl& vhost,
+                                         const envoy::api::v2::Route& route,
+                                         Runtime::Loader& loader)
+    : RouteEntryImplBase(vhost, route, loader),
+      regex_(std::regex{route.match().regex(), std::regex::optimize}) {}
+
+void RegexRouteEntryImpl::finalizeRequestHeaders(Http::HeaderMap& headers) const {
+  RouteEntryImplBase::finalizeRequestHeaders(headers);
+
+  const Http::HeaderString& path = headers.Path()->value();
+  const char* query_string_start = findQueryStringStart(path);
+  ASSERT(std::regex_match(path.c_str(), query_string_start, regex_));
```

Comment:
dumb question: what is this assert verifying?

Reply:
It asserts that the path in `headers` matches the regex, before going ahead and replacing the path with `prefix_rewrite` (which happens in `finalizePathHeader`).

Reply:
The reason I added the assert is that `finalizePathHeader` asserts that the caller knows what it's replacing before it swaps in `prefix_rewrite`. This made sense for prefix and path matching, but I have to artificially feed in `matched_path` for regex matching.

Reply:
Ah I see. Thanks for the explanation.